File: app/javascript/src/queries/chatQueries.js
import gql from "graphql-tag";

export const GET_CHATS = gql`
  query Chats($offset: Int!, $limit: Int!) {
    chats(offset: $offset, limit: $limit)
      @connection(key: "messages", filter: ["chatId"])
      @connection(key: "chats") {
      id
      isUnread
      lastMessage {
        id
        senderId
        body
        createdAt
      }
      contact {
        id
        slug
        name
        avatar
      }
    }
  }
`;

export const GET_MESSAGES = gql`
  query Messages($caseId: ID, $caseType: String, $chatId: ID, $offset: Int!, $limit: Int!) {
    messages(caseId: $caseId, caseType: $caseType, chatId: $chatId, offset: $offset, limit: $limit)
      @connection(key: "messages", filter: ["chatId"]) {
      id
      body
      senderId
      createdAt
    }
  }
`;

export const GET_UNREAD_CHATS_COUNT = gql`
  query UnreadChatsCount {
    unreadChatsCount
  }
`;

File: app/graphql/resolvers/queries/chat_queries.rb
module Resolvers
    module Queries
        module ChatQueries
            class GetChats < GraphQL::Function
                type types[Types::ChatType]

                argument :offset, !types.Int
                argument :limit, !types.Int

                def call(obj, args, ctx)
                    Chat.where(recipient: ctx[:current_user]).includes(:last_message, :sender, :recipient)
                end
            end
            
            class GetModerationMessages < GraphQL::Function
                type types[Types::MessageType]

                argument :caseId, types.ID
                argument :chatId, types.ID
                argument :caseType, types.String
                argument :offset, !types.Int
                argument :limit, !types.Int

                def call(obj, args, ctx)
                    if args[:chatId].present?
                        Message.where(chat: Chat.find(args[:chatId])).order("messages.created_at ASC")
                    else 
                        Message.where(chat: Chat.find_by(case_id: args[:caseId], case_type: args[:caseType])).order("messages.created_at ASC")
                    end
                end
            end
            
            class GetMessages < GraphQL::Function
                type types[Types::MessageType]

                argument :chatId, types.ID
                argument :offset, !types.Int
                argument :limit, !types.Int

                def call(obj, args, ctx)
                    MessagePolicy::Scope.new(ctx[:current_user], Message.where(chat_id: args[:chatId])).resolve.order("messages.created_at ASC")#.offset(args[:offset]).limit(args[:limit])
                end
            end

        end
    end
end
